### PR TITLE
nested op_name scope in reduce functions

### PR DESCRIPTION
**Pull request by gustafsson**
This commit makes nested op_name scopes work for the reduce_-functions.

### CODE REVIEW

**Comment by oxinabox** [comment]
This has broken some tests.
https://travis-ci.org/malmaud/TensorFlow.jl/jobs/353723466#L1184

I've not looked too closely as to why,
Might the the tests are too fragile, might not.

**Comment by gustafsson** [comment]
Thanks, I updated the failing test.

Since reduce without axis is implemented with multiple nodes I think that `@tf Ysum1 = reduce_sum(Y)` should create a scope that contains these nodes. Since the last of those nodes is a sum the output node is then named `Ysum1/sum`.

If an axis is given explicitly `Ysum5 = reduce_sum(Y, axis=2)` there's just one node, and it can then take the name Ysum5.

**Comment by oxinabox** [comment]
It bothers me that we have a function that takes a `name` argument and returns a Node.
and even  if we provide the name argument directly, the returned node does not have that name.

I'm not sure what the resolution is.

This is definitely a breaking change -- it will break some of my code I am certain.

**Comment by gustafsson** [comment]
Ah, just because there is a node named `reduce_sum/rank` doesn't prevent us from creating a node named `reduce_sum`. I.e that's kind of what this function did before this PR.

However before this PR the node was named `reduce`, is that inherited from python? I think `reduce_sum` (i.e `reduce_$reduction`) or would be a better name.

**Comment by oxinabox** [comment]
Hmm,
Can you pop out some python graphs and compare?

**Comment by gustafsson** [comment]
Here are some graphs. tensorflow.py v1.0 and v1.8 were near identical.

TensorFlow.jl known rank:
reduce-known_rank-tensorflow jl

tensorflow.py v1.8 known rank:
reduce-known_rank-tensorflow-v1 8

TensorFlow.jl unknown rank:
reduce-unknown_rank-tensorflow jl

tensorflow.py v1.8 unknown rank:
reduce-unknown_rank-tensorflow-v1 8

**Comment by gustafsson** [comment]
Code for the above:

julia
```
using TensorFlow; tf = TensorFlow

v = tf.placeholder(Float32, shape=[2,3,5])

myop = tf.with_op_name("myop") do
    tf.reduce_sum(v) + tf.reduce_sum(v) + tf.reduce_sum(v, name="s")
end

myop_dim = tf.with_op_name("myop3") do
    tf.reduce_sum(v, axis=3) + tf.reduce_sum(v, axis=3) + tf.reduce_sum(v, axis=3, name="dims")
end

myop + myop_dim

sess = tf.Session()
fw = tf.summary.FileWriter("models/tensorflow.jl")
```

python
```
import tensorflow as tf

v = tf.placeholder(tf.float32, shape=[2,3,4])

with(tf.name_scope("myop")):
    myop = tf.reduce_sum(v) + tf.reduce_sum(v) + tf.reduce_sum(v, name='s')

with(tf.name_scope("myop2")):
    myop2 = tf.reduce_sum(v, 2) + tf.reduce_sum(v, 2) + tf.reduce_sum(v, 2, name="dims")

myop + myop2

sess = tf.Session()
fw = tf.summary.FileWriter('models/tensorflow.py')
fw.add_graph(sess.graph)
```

**Comment by oxinabox** [comment]
This looks good to me.
@malmaud  can you give this a once over?

**Comment by malmaud** [comment]
Yes, someone this weekend.

**Comment by malmaud** [comment]
LGTM. Seems to be a a conflict now though.

**Comment by malmaud** [comment]
@gustafsson Are you interested in rebasing this on the current TensorFlow.jl master?
